teps/0099-parameters-in-script.md

```diff
@@ -0,0 +1,479 @@
+
+This was the proposal in [TEP-0017][tep-0017].
+
+> In a Task or Pipeline, for each parameter `foo`, provide `$(params.foo.shell-escaped)`.
+>
+> This variable shall be escaped according to `printf '%q'` rules as used in bash.
+> For example, the literal string `Hello '\!"` becomes `Hello\ \'\\\!\"`. A script
+> which ran `echo $(params.foo.shell-escaped)` would print the literal string. The
+> escaped value would do the wrong thing when placed inside double or single quotes.
+
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.bar)
+  
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.bar.shell-escaped)
+```
+
+##### Discussion
+
+This approach mutates the specification, but it is something we're already doing
+with implicit `Parameters` per [TEP-0023][tep-0023].
+
+### 4. Reject `Parameters` in `Script` and provide instructions to fix
+
+During validation we could detect `Parameters` used in `Script` blocks and return
+a helpful message to the user that they need to rewrite their `Task` to move the
+`Parameters` into environment variables.
+
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.bar)
+```
+
+When the Tekton Pipelines Controller sees this `Script` it rejects the `Task` submission
+with an error, such as:
+
+```
+spec.steps[0]: params cannot be referenced from a script, consider moving
+into an environment variable on the Step.
+```
+
+```yaml      
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    env:
+    - name: PARAM_BAR
+      value: $(params.bar)
+    script: |
+      echo ${PARAM_BAR}      
+```
+
+##### Discussion
+
+Non-mutating but doesn't help the user as directly as mutating the spec.
+
+### 5. Optionally Project `Parameters` as Environment Variables in `Steps`
+
+Building on the first alternative above, another approach would be to make
+the implicit projection optional, where it only applies to a subset of
+`Parameters` with the suffix `env`.
+
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.bar)
+  
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.bar.env)
+      
+  # After - Resolved
+  steps:
+  - name: foo
+    image: myimage
+    env:
+    - name: PARAM_BAR
+      value: $(params.bar)
+    script: |
+      echo ${PARAM_BAR}
+```
+
+##### Discussion
+
+This approach is flexible: provide a safer and more reliable interpolation while
+still allowing code injection if needed. Even more, it is backwards compatible.
+However, the string replacement in `Script` would only work for shell scripts .
+
+## Open Questions
+
+1. Is it sufficient to address `Parameters` in `Script` only, or do we need to
+   consider all variables that can be substituted in `Script`?
+1. Which alternative do we prefer to go forward with?
```

Comment:
I definitely prefer the approach of projecting into env or filesystem rather than shell-escaping the value. Shell-escaping feels language-specific while script fields are language agnostic. I also like the explicitness of using `$(params.foo.env)` or `$(params.foo.path)` in the script over implicit projections.

Exposing params exclusively via the filesystem makes the most sense to me if we decide to take that route in TEP-0086 for populating params from a sidecar.

Reply:
Agreed with concerns about shell-escaping. My preferred approach would be to disallow interpolation in scripts, and to have an opt-out feature that makes parameters available as env vars (as proposed in alternative #1).

Reply:
I've mentioned my objections to shell escaping, and think it's a non-starter.  To me, projecting to a file system would be nice-to-have, as some parameters might contain (important) newlines, special characters and so on.  However, having to read in each parameter to a variable can be tedious and verbose, and easily fumbled, so I would much much prefer the automatic environment variable mapping as the default.

```
params:
  - name: branch
    description: The name of a branch (no whitespace or trailing newline allowed)
    type: string
    # projection: environment  # the default perhaps?
  - name: manifest
    description: A yaml manifest (that should end with a trailing newline)
    type: string
    projection: file  # or something.
```

This would supply 
* branch as the PARAM_BRANCH environment variable
* manifest as /tekton/params/manifest, and its path provided as $(params.manifest.path) perhaps.

The "projection" could in future be used for an as-of-yet unused feature: the script's standard input!  e.g. `projection: stdin` = pass a parameter to the stdin of the command.  That could allow you to have `sed` or `jq` as the shell, to do simple transformations of data as tasks.  It would of course be awesome if that is paired with the script's stdout being piped to a _result_.